ct-ng build fails to link libduma for arm-unknown-linux-uclibcgnueabi (Ubuntu 16.04 32-bit )

The effect of running `ct-ng build` to create a toolchain from the sample `arm-unknown-linux-uclibcgnueabi` is a failure during the step "Installing D.U.M.A.". I've posted the tail end of the log below and I have attached a link to the complete log as well (it's really long, sorry).

I changed the following settings from the given sample:

- Parallel jobs = 8
- [*] Build Static Toolchain
- Linux Kernel Version (3.2.77) -- this is closest to the kernel on the target
- [] Build Shared Libraries
- gcc version (4.9.3) -- The problem is the same with version 5.4.0

The host is Ubuntu 16.04 32-bit in VirtualBox.

```
[ALL  ]    arm-unknown-linux-uclibcgnueabi-gcc -g -O0 -Wall -Wextra -fPIC -DPIC -DDUMA_SO_LIBRARY -c dumapp.cpp -o dumapp_so.o
[ALL  ]    arm-unknown-linux-uclibcgnueabi-gcc -g -shared -Wl,-soname,libduma.so.0.0.0 -o libduma.so.0.0.0 duma_so.o sem_inc_so.o print_so.o dumapp_so.o -lpthread -lc
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/arm-unknown-linux-uclibcgnueabi/sysroot/usr/lib/libc.a(memcpy.os): In function `__GI_memcpy':
[ALL  ]    (.text+0x0): multiple definition of `memcpy'
[ALL  ]    duma_so.o:/home/farhan/tc/arm-unknown-linux-uclibcgnueabi/.build/arm-unknown-linux-uclibcgnueabi/build/build-duma/duma.c:2443: first defined here
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/arm-unknown-linux-uclibcgnueabi/sysroot/usr/lib/libc.a(strcpy.os): In function `strcpy':
[ALL  ]    strcpy.c:(.text+0x0): multiple definition of `strcpy'
[ALL  ]    duma_so.o:/home/farhan/tc/arm-unknown-linux-uclibcgnueabi/.build/arm-unknown-linux-uclibcgnueabi/build/build-duma/duma.c:2449: first defined here
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/arm-unknown-linux-uclibcgnueabi/sysroot/usr/lib/libc.a(__uClibc_main.os): In function `__uClibc_fini':
[ALL  ]    __uClibc_main.c:(.text+0x12c): undefined reference to `__fini_array_end'
[ALL  ]    __uClibc_main.c:(.text+0x130): undefined reference to `__fini_array_start'
[ALL  ]    __uClibc_main.c:(.text+0x134): undefined reference to `__fini_array_start'
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/arm-unknown-linux-uclibcgnueabi/sysroot/usr/lib/libc.a(__uClibc_main.os): In function `__uClibc_main':
[ALL  ]    __uClibc_main.c:(.text+0x55c): undefined reference to `__preinit_array_start'
[ALL  ]    __uClibc_main.c:(.text+0x560): undefined reference to `__preinit_array_end'
[ALL  ]    __uClibc_main.c:(.text+0x564): undefined reference to `__preinit_array_start'
[ALL  ]    __uClibc_main.c:(.text+0x568): undefined reference to `__init_array_start'
[ALL  ]    __uClibc_main.c:(.text+0x56c): undefined reference to `__init_array_end'
[ALL  ]    __uClibc_main.c:(.text+0x570): undefined reference to `__init_array_start'
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/lib/gcc/arm-unknown-linux-uclibcgnueabi/4.9.3/../../../../arm-unknown-linux-uclibcgnueabi/bin/ld: libduma.so.0.0.0: hidden symbol `__fini_array_end' isn't defined
[ALL  ]    /home/farhan/x-tools/arm-unknown-linux-uclibcgnueabi/lib/gcc/arm-unknown-linux-uclibcgnueabi/4.9.3/../../../../arm-unknown-linux-uclibcgnueabi/bin/ld: final link failed: Bad value
[ERROR]    collect2: error: ld returned 1 exit status
[ALL  ]    GNUmakefile:523: recipe for target 'libduma.so.0.0.0' failed
[ERROR]    make[1]: *** [libduma.so.0.0.0] Error 1
[ALL  ]    make[1]: Leaving directory '/home/farhan/tc/arm-unknown-linux-uclibcgnueabi/.build/arm-unknown-linux-uclibcgnueabi/build/build-duma'
[ERROR]  
[ERROR]  >>
[ERROR]  >>  Build failed in step 'Installing D.U.M.A.'
[ERROR]  >>        called in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@259]
[ERROR]  >>        called from: do_debug_duma_build[scripts/build/debug/200-duma.sh@42]
[ERROR]  >>        called from: do_debug[scripts/build/debug.sh@35]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@624]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: 'build.log'
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      '/home/farhan/.local/share/doc/crosstool-ng/crosstool-ng-1.22.0-217-g1e4040e/B - Known issues.txt'
[ERROR]  
[ERROR]  (elapsed: 172:58.74)
```

Thanks for your time and work. If I can get this all to work, I can go ahead and make some PRs to add some to the documentation (README.md) with what I've learned as I used this software (dependencies, and how to install them, configuration options, etc) as a thank you.

[Here is the full build.log of the failed build](https://gist.github.com/kathawala/74ecb31e7bb59a39b362af7ca7494bf5)

One additional problem though, with the build configuration above (static toolchain, no shared libs) "native gdb" will fail due to threads && shared library dependency (log below). If thread support is present "native gdb" will also need shared library support I guess. 

```
[INFO ]  Performing some trivial sanity checks
[INFO ]  Build started 20160907.163934
[INFO ]  Building environment variables
[WARN ]  Directory '/home/user/src' does not exist.
[WARN ]  Will not save downloaded tarballs to local storage.
[EXTRA]  Preparing working directories
[EXTRA]  Installing user-supplied crosstool-NG configuration
[EXTRA]  =================================================================
[EXTRA]  Dumping internal crosstool-NG configuration
[EXTRA]    Building a toolchain for:
[EXTRA]      build  = x86_64-pc-linux-gnu
[EXTRA]      host   = x86_64-pc-linux-gnu
[EXTRA]      target = arm-unknown-linux-uclibcgnueabi
[EXTRA]  Dumping internal crosstool-NG configuration: done in 0.03s (at 00:01)
[INFO ]  =================================================================
[INFO ]  Retrieving needed toolchain components' tarballs
[INFO ]  Retrieving needed toolchain components' tarballs: done in 0.05s (at 00:01)
[INFO ]  =================================================================
[INFO ]  Extracting and patching toolchain components
[EXTRA]    Extracting 'linux-3.2.77'
[EXTRA]    Patching 'linux-3.2.77'
[EXTRA]    Extracting 'gmp-6.1.0'
[EXTRA]    Patching 'gmp-6.1.0'
[EXTRA]    Extracting 'mpfr-3.1.3'
[EXTRA]    Patching 'mpfr-3.1.3'
[EXTRA]    Extracting 'isl-0.14'
[EXTRA]    Patching 'isl-0.14'
[EXTRA]    Extracting 'cloog-0.18.4'
[EXTRA]    Patching 'cloog-0.18.4'
[EXTRA]    Extracting 'mpc-1.0.3'
[EXTRA]    Patching 'mpc-1.0.3'
[EXTRA]    Extracting 'libelf-0.8.13'
[EXTRA]    Patching 'libelf-0.8.13'
[EXTRA]    Extracting 'expat-2.1.0'
[EXTRA]    Patching 'expat-2.1.0'
[EXTRA]    Extracting 'ncurses-6.0'
[EXTRA]    Patching 'ncurses-6.0'
[EXTRA]    Extracting 'binutils-2.26'
[EXTRA]    Patching 'binutils-2.26'
[EXTRA]    Extracting 'gcc-4.9.3'
[EXTRA]    Patching 'gcc-4.9.3'
[EXTRA]    Extracting 'uClibc-ng-1.0.12'
[EXTRA]    Patching 'uClibc-ng-1.0.12'
[EXTRA]    Extracting 'dmalloc-5.5.2'
[EXTRA]    Patching 'dmalloc-5.5.2'
[EXTRA]    Extracting 'duma_2_5_15'
[EXTRA]    Patching 'duma-2_5_15'
[EXTRA]    Extracting 'gdb-7.11'
[EXTRA]    Patching 'gdb-7.11'
[EXTRA]    Extracting 'ltrace-0.7.3'
[EXTRA]    Patching 'ltrace-0.7.3'
[EXTRA]    Extracting 'strace-4.10'
[EXTRA]    Patching 'strace-4.10'
[INFO ]  Extracting and patching toolchain components: done in 31.18s (at 00:32)
[INFO ]  =================================================================
[INFO ]  Installing ncurses for build
[EXTRA]    Configuring ncurses
[EXTRA]    Building ncurses
[EXTRA]    Installing ncurses
[INFO ]  Installing ncurses for build: done in 20.90s (at 00:53)
[INFO ]  =================================================================
[INFO ]  Installing GMP for host
[EXTRA]    Configuring GMP
[EXTRA]    Building GMP
[EXTRA]    Installing GMP
[INFO ]  Installing GMP for host: done in 28.05s (at 01:21)
[INFO ]  =================================================================
[INFO ]  Installing MPFR for host
[EXTRA]    Configuring MPFR
[EXTRA]    Building MPFR
[EXTRA]    Installing MPFR
[INFO ]  Installing MPFR for host: done in 15.68s (at 01:37)
[INFO ]  =================================================================
[INFO ]  Installing ISL for host
[EXTRA]    Configuring ISL
[EXTRA]    Building ISL
[EXTRA]    Installing ISL
[INFO ]  Installing ISL for host: done in 13.49s (at 01:50)
[INFO ]  =================================================================
[INFO ]  Installing CLooG for host
[EXTRA]    Configuring CLooG
[EXTRA]    Building CLooG
[EXTRA]    Installing CLooG
[INFO ]  Installing CLooG for host: done in 3.75s (at 01:54)
[INFO ]  =================================================================
[INFO ]  Installing MPC for host
[EXTRA]    Configuring MPC
[EXTRA]    Building MPC
[EXTRA]    Installing MPC
[INFO ]  Installing MPC for host: done in 6.72s (at 02:01)
[INFO ]  =================================================================
[INFO ]  Installing expat for host
[EXTRA]    Configuring expat
[EXTRA]    Building expat
[EXTRA]    Installing expat
[INFO ]  Installing expat for host: done in 4.07s (at 02:05)
[INFO ]  =================================================================
[INFO ]  Installing binutils for host
[EXTRA]    Configuring binutils
[EXTRA]    Prepare binutils for static build
[EXTRA]    Building binutils
[EXTRA]    Installing binutils
[INFO ]  Installing binutils for host: done in 39.75s (at 02:45)
[INFO ]  =================================================================
[INFO ]  Installing pass-1 core C gcc compiler
[EXTRA]    Configuring core C gcc compiler
[EXTRA]    Building gcc
[EXTRA]    Installing gcc
[EXTRA]    Housekeeping for core gcc compiler
[EXTRA]    gcc default flags: ''
[EXTRA]    gcc configured with these multilibs (including the default):
[EXTRA]       '' --> lib (gcc)   lib (os)
[INFO ]  Installing pass-1 core C gcc compiler: done in 155.91s (at 05:21)
[INFO ]  =================================================================
[INFO ]  Installing kernel headers
[EXTRA]    Installing kernel headers
[EXTRA]    Checking installed headers
[INFO ]  Installing kernel headers: done in 3.65s (at 05:24)
[INFO ]  =================================================================
[INFO ]  Installing C library headers & start files
[INFO ]    =================================================================
[INFO ]    Building for multilib 1/1: ''
[EXTRA]      Copying sources to build dir
[EXTRA]      Applying configuration
[EXTRA]      Building headers
[EXTRA]      Installing headers
[EXTRA]      Building start files
[EXTRA]      Building dummy shared libs
[EXTRA]      Installing start files
[EXTRA]      Installing dummy shared libs
[INFO ]    Building for multilib 1/1: '': done in 6.64s (at 05:31)
[INFO ]  Installing C library headers & start files: done in 6.68s (at 05:31)
[INFO ]  =================================================================
[INFO ]  Installing pass-2 core C gcc compiler
[EXTRA]    Configuring core C gcc compiler
[EXTRA]    Building gcc
[EXTRA]    Installing gcc
[EXTRA]    Housekeeping for core gcc compiler
[EXTRA]    gcc default flags: ''
[EXTRA]    gcc configured with these multilibs (including the default):
[EXTRA]       '' --> lib (gcc)   lib (os)
[INFO ]  Installing pass-2 core C gcc compiler: done in 229.50s (at 09:20)
[INFO ]  =================================================================
[INFO ]  Installing C library
[INFO ]    =================================================================
[INFO ]    Building for multilib 1/1: ''
[EXTRA]      Copying sources to build dir
[EXTRA]      Applying configuration
[EXTRA]      Cleaning up startfiles
[EXTRA]      Building C library
[EXTRA]      Installing C library
[INFO ]    Building for multilib 1/1: '': done in 24.15s (at 09:45)
[INFO ]  Installing C library: done in 24.19s (at 09:45)
[INFO ]  =================================================================
[INFO ]  Installing final gcc compiler
[EXTRA]    Configuring final gcc compiler
[EXTRA]    Building final gcc compiler
[EXTRA]    Installing final gcc compiler
[EXTRA]    Housekeeping for final gcc compiler
[EXTRA]    gcc default flags: ''
[EXTRA]    gcc configured with these multilibs (including the default):
[EXTRA]       '' --> lib (gcc)   lib (os)
[INFO ]  Installing final gcc compiler: done in 236.37s (at 13:41)
[INFO ]  =================================================================
[INFO ]  Checking dynamic linker symlinks
[EXTRA]    Checking dynamic linker for multilib ''
[INFO ]  Checking dynamic linker symlinks: done in 0.07s (at 13:41)
[INFO ]  =================================================================
[INFO ]  Installing libelf for the target
[EXTRA]    Configuring libelf
[EXTRA]    Building libelf
[EXTRA]    Installing libelf
[INFO ]  Installing libelf for the target: done in 2.65s (at 13:44)
[INFO ]  =================================================================
[INFO ]  Installing expat for target
[EXTRA]    Configuring expat
[EXTRA]    Building expat
[EXTRA]    Installing expat
[INFO ]  Installing expat for target: done in 2.74s (at 13:47)
[INFO ]  =================================================================
[INFO ]  Installing ncurses for target
[EXTRA]    Configuring ncurses
[EXTRA]    Building ncurses
[EXTRA]    Installing ncurses
[INFO ]  Installing ncurses for target: done in 18.53s (at 14:05)
[INFO ]  =================================================================
[INFO ]  Installing dmalloc
[EXTRA]    Configuring dmalloc
[EXTRA]    Building dmalloc
[EXTRA]    Installing dmalloc
[INFO ]  Installing dmalloc: done in 4.78s (at 14:10)
[INFO ]  =================================================================
[INFO ]  Installing D.U.M.A.
[EXTRA]    Copying sources
[EXTRA]    Building libraries 'libduma.a'
[EXTRA]    Installing libraries 'libduma.a'
[INFO ]  Installing D.U.M.A.: done in 0.42s (at 14:10)
[INFO ]  =================================================================
[INFO ]  Installing cross-gdb
[EXTRA]    Configuring cross-gdb
[EXTRA]    Building cross-gdb
[EXTRA]    Installing cross-gdb
[EXTRA]    Installing '.gdbinit' template
[INFO ]  Installing cross-gdb: done in 88.29s (at 15:39)
[INFO ]  =================================================================
[INFO ]  Installing native gdb
[EXTRA]    Configuring native gdb
[EXTRA]    Building native gdb
[ERROR]    /opt/build/.build/src/gdb-7.11/gdb/linux-thread-db.c:21:19: fatal error: dlfcn.h: No such file or directory
[ERROR]    make[3]: *** [linux-thread-db.o] Error 1
[ERROR]    make[2]: *** [all-gdb] Error 2
[ERROR]    make[1]: *** [all] Error 2
[ERROR]   
[ERROR]  >>
[ERROR]  >>  Build failed in step 'Installing native gdb'
[ERROR]  >>        called in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoExecLog[scripts/functions@259]
[ERROR]  >>        called from: do_debug_gdb_build[scripts/build/debug/300-gdb.sh@233]
[ERROR]  >>        called from: do_debug[scripts/build/debug.sh@35]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@624]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: 'build.log'
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      '/opt/ct/share/doc/crosstool-ng/crosstool-ng-1.22.0-217-g1e4040e-dirty/B - Known issues.txt'
[ERROR]   
[ERROR]  (elapsed: 16:35.15)
[16:35] / make: *** [build] Error 2
```

--- a/config/debug/duma.in
+++ b/config/debug/duma.in
@@ -14,7 +14,8 @@ config DUMA_A
 config DUMA_SO
     bool
     prompt "Build a shared library"
-    default y if SHARED_LIBS
+    depends on SHARED_LIBS
+    default y
 
 choice
     bool

